File: app/context/materials.py
"""会前资料加载：摘要 / PPT 文本 / 简介 / 术语表 → 纯文本（STATIC 上下文）。"""
from __future__ import annotations

import json
from pathlib import Path


def load_materials(materials_dir: Path, max_chars: int = 4000) -> str:
    """读取 materials 目录下所有可解析文件，拼接为纯文本。"""
    if not materials_dir.exists():
        return ""
    chunks: list[str] = []
    for f in sorted(materials_dir.iterdir()):
        if not f.is_file():
            continue
        try:
            text = _load_file(f)
        except Exception:
            continue
        if text:
            chunks.append(f"[{f.stem}]\n{text.strip()}")
        if sum(len(c) for c in chunks) >= max_chars:
            break
    joined = "\n\n".join(chunks)
    return joined[:max_chars]


def _load_file(f: Path) -> str:
    suffix = f.suffix.lower()
    if suffix in (".md", ".txt", ".text"):
        return f.read_text(encoding="utf-8", errors="ignore")
    if suffix == ".json":
        data = json.loads(f.read_text(encoding="utf-8", errors="ignore"))
        return json.dumps(data, ensure_ascii=False, indent=1)
    # PDF / PPTX 等：MVP 阶段给出提示（可选安装依赖后启用）
    return ""

File: go/internal/contextx/materials.go
// Package contextx 上下文组装：会前资料 / 最近转写 / 焦点 / 历史。
// 对应 Python 版 app/context/（包名用 contextx 以避免与标准库 context 冲突）。
package contextx

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const defaultMaterialsMaxChars = 4000

// LoadMaterials 读取 materials 目录下所有可解析文件，拼接为纯文本（STATIC 上下文）。
func LoadMaterials(materialsDir string, maxChars int) string {
	if maxChars <= 0 {
		maxChars = defaultMaterialsMaxChars
	}
	entries, err := os.ReadDir(materialsDir)
	if err != nil {
		return ""
	}
	// Python 版用 sorted(iterdir())：按文件名字典序
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })

	var chunks []string
	total := 0
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		text := loadFile(filepath.Join(materialsDir, e.Name()))
		if text != "" {
			stem := strings.TrimSuffix(e.Name(), filepath.Ext(e.Name()))
			chunk := "[" + stem + "]\n" + strings.TrimSpace(text)
			chunks = append(chunks, chunk)
			total += len([]rune(chunk))
		}
		if total >= maxChars {
			break
		}
	}
	joined := strings.Join(chunks, "\n\n")
	r := []rune(joined)
	if len(r) > maxChars {
		return string(r[:maxChars])
	}
	return joined
}

// loadFile 按后缀解析单个资料文件；PDF/PPTX 等暂不支持（返回空串）。
func loadFile(path string) string {
	raw, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	switch strings.ToLower(filepath.Ext(path)) {
	case ".md", ".txt", ".text":
		return string(raw)
	case ".json":
		var data any
		if err := json.Unmarshal(raw, &data); err != nil {
			return string(raw)
		}
		b, err := json.MarshalIndent(data, "", " ")
		if err != nil {
			return string(raw)
		}
		return string(b)
	default:
		return ""
	}
}
